File: Day52/Part5/hex_arena.py
from collections import deque

def solve_hex_arena(hex_list):
    # hex_list: list of (q, r, char)
    grid = {}
    start_pos = None
    end_pos = None
    
    for q, r, char in hex_list:
        grid[(q, r)] = char
        if char == 'S':
            start_pos = (q, r)
        elif char == 'E':
            end_pos = (q, r)
            
    if not start_pos or not end_pos:
        return -1
        
    # BFS
    # State: (q, r, mask)
    queue = deque([(start_pos[0], start_pos[1], 0, 0)]) # q, r, mask, steps
    visited = set([(start_pos[0], start_pos[1], 0)])
    
    # Hex neighbors (Axial)
    # +q, -r | +q, 0 | 0, +r | -q, +r | -q, 0 | 0, -r ??
    # Standard Axial neighbors:
    # (1, 0), (1, -1), (0, -1), (-1, 0), (-1, 1), (0, 1)
    neighbors = [
        (1, 0), (1, -1), (0, -1),
        (-1, 0), (-1, 1), (0, 1)
    ]
    
    while queue:
        q, r, mask, steps = queue.popleft()
        
        if (q, r) == end_pos:
            return steps
            
        for dq, dr in neighbors:
            nq, nr = q + dq, r + dr
            
            if (nq, nr) in grid:
                cell = grid[(nq, nr)]
                
                if cell == '#':
                    continue
                    
                new_mask = mask
                is_passable = True
                
                # Check Barrier
                if 'A' <= cell <= 'J' and cell != 'E':
                    barrier_id = ord(cell) - ord('A')
                    if not (mask & (1 << barrier_id)):
                        is_passable = False
                        
                # Check Plate
                if 'a' <= cell <= 'j':
                    plate_id = ord(cell) - ord('a')
                    new_mask = mask ^ (1 << plate_id)
                    
                if is_passable:
                    if (nq, nr, new_mask) not in visited:
                        visited.add((nq, nr, new_mask))
                        queue.append((nq, nr, new_mask, steps + 1))
                        
    return -1

if __name__ == "__main__":
    # Example 1
    # S -> a -> A -> E
    # (0,0) S
    # (1,0) a (Plate 0)
    # (2,0) A (Barrier 0)
    # (3,0) E
    
    hexes = [
        (0, 0, 'S'),
        (1, 0, 'a'),
        (2, 0, 'A'),
        (3, 0, 'E')
    ]
    print(f"Min Steps (Ex 1): {solve_hex_arena(hexes)}")
    
    # Example 2: Blocked
    # S -> A -> E (No switch)
    hexes2 = [
        (0, 0, 'S'),
        (1, 0, 'A'),
        (2, 0, 'E')
    ]
    print(f"Min Steps (Ex 2): {solve_hex_arena(hexes2)}")

File: Day52/Part5/hex_arena.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <map>
#include <set>
#include <tuple>

using namespace std;

struct Hex {
    int q, r;
    char type;
};

struct State {
    int q, r;
    int mask;
    int steps;
};

// For set/map keys
using Coord = pair<int, int>;
using VisitedState = tuple<int, int, int>; // q, r, mask

int solve_hex_arena(const vector<Hex>& hex_list) {
    map<Coord, char> grid;
    int start_q, start_r;
    int end_q, end_r;
    bool found_start = false, found_end = false;

    for (const auto& h : hex_list) {
        grid[{h.q, h.r}] = h.type;
        if (h.type == 'S') {
            start_q = h.q;
            start_r = h.r;
            found_start = true;
        } else if (h.type == 'E') {
            end_q = h.q;
            end_r = h.r;
            found_end = true;
        }
    }

    if (!found_start || !found_end) return -1;

    queue<State> q;
    q.push({start_q, start_r, 0, 0});

    set<VisitedState> visited;
    visited.insert({start_q, start_r, 0});

    // Axial neighbors
    int dq[] = {1, 1, 0, -1, -1, 0};
    int dr[] = {0, -1, -1, 0, 1, 1};

    while (!q.empty()) {
        State current = q.front();
        q.pop();

        if (current.q == end_q && current.r == end_r) {
            return current.steps;
        }

        for (int i = 0; i < 6; ++i) {
            int nq = current.q + dq[i];
            int nr = current.r + dr[i];

            if (grid.find({nq, nr}) == grid.end()) continue;

            char cell = grid[{nq, nr}];
            if (cell == '#') continue;

            int new_mask = current.mask;
            bool is_passable = true;

            // Check Barrier (A-J), excluding E
            if (cell >= 'A' && cell <= 'J' && cell != 'E') {
                int barrier_id = cell - 'A';
                if (!(current.mask & (1 << barrier_id))) {
                    is_passable = false;
                }
            }

            // Check Plate (a-j)
            if (cell >= 'a' && cell <= 'j') {
                int plate_id = cell - 'a';
                new_mask = current.mask ^ (1 << plate_id);
            }

            if (is_passable) {
                if (visited.find({nq, nr, new_mask}) == visited.end()) {
                    visited.insert({nq, nr, new_mask});
                    q.push({nq, nr, new_mask, current.steps + 1});
                }
            }
        }
    }

    return -1;
}

int main() {
    // Example 1
    vector<Hex> hexes1 = {
        {0, 0, 'S'},
        {1, 0, 'a'},
        {2, 0, 'A'},
        {3, 0, 'E'}
    };
    cout << "Min Steps (Ex 1): " << solve_hex_arena(hexes1) << endl;

    // Example 2
    vector<Hex> hexes2 = {
        {0, 0, 'S'},
        {1, 0, 'A'},
        {2, 0, 'E'}
    };
    cout << "Min Steps (Ex 2): " << solve_hex_arena(hexes2) << endl;

    return 0;
}
